gcloud-java-storage/src/test/java/com/google/gcloud/storage/ITStorageTest.java

```diff
@@ -0,0 +1,470 @@
+    BlobInfo remoteBlob = storage.create(blob, BLOB_BYTE_CONTENT);
+    assertNotNull(remoteBlob);
+    assertEquals(blob.bucket(), remoteBlob.bucket());
+    assertEquals(blob.name(), remoteBlob.name());
+    byte[] readBytes = storage.readAllBytes(bucket, blobName);
+    assertArrayEquals(BLOB_BYTE_CONTENT, readBytes);
+    assertTrue(storage.delete(bucket, blobName));
+  }
+
+  @Test
+  public void testCreateEmptyBlob() {
+    String blobName = "test-create-empty-blob";
+    BlobInfo blob = BlobInfo.of(bucket, blobName);
+    BlobInfo remoteBlob = storage.create(blob);
+    assertNotNull(remoteBlob);
+    assertEquals(blob.bucket(), remoteBlob.bucket());
+    assertEquals(blob.name(), remoteBlob.name());
+    byte[] readBytes = storage.readAllBytes(bucket, blobName);
+    assertArrayEquals(new byte[0], readBytes);
+    assertTrue(storage.delete(bucket, blobName));
+  }
+
+  @Test
+  public void testCreateBlobStream() throws UnsupportedEncodingException {
+    String blobName = "test-create-blob-stream";
+    BlobInfo blob = BlobInfo.builder(bucket, blobName).contentType(CONTENT_TYPE).build();
+    ByteArrayInputStream stream = new ByteArrayInputStream(BLOB_STRING_CONTENT.getBytes(UTF_8));
+    BlobInfo remoteBlob = storage.create(blob, stream);
+    assertNotNull(remoteBlob);
+    assertEquals(blob.bucket(), remoteBlob.bucket());
+    assertEquals(blob.name(), remoteBlob.name());
+    assertEquals(blob.contentType(), remoteBlob.contentType());
+    byte[] readBytes = storage.readAllBytes(bucket, blobName);
+    assertEquals(BLOB_STRING_CONTENT, new String(readBytes, UTF_8));
+    assertTrue(storage.delete(bucket, blobName));
+  }
+
+  @Test
+  public void testCreateBlobFail() {
+    String blobName = "test-create-blob-fail";
+    BlobInfo blob = BlobInfo.of(bucket, blobName);
+    assertNotNull(storage.create(blob));
+    try {
+      storage.create(blob.toBuilder().generation(42L).build(), BLOB_BYTE_CONTENT,
+          Storage.BlobTargetOption.generationMatch());
+      fail("StorageException was expected");
+    } catch (StorageException ex) {
+      // expected
+    }
+    assertTrue(storage.delete(bucket, blobName));
+  }
+
+  @Test
+  public void testUpdateBlob() {
+    String blobName = "test-update-blob";
+    BlobInfo blob = BlobInfo.of(bucket, blobName);
+    assertNotNull(storage.create(blob));
+    BlobInfo updatedBlob = storage.update(blob.toBuilder().contentType(CONTENT_TYPE).build());
+    assertNotNull(updatedBlob);
+    assertEquals(blob.bucket(), updatedBlob.bucket());
+    assertEquals(blob.name(), updatedBlob.name());
+    assertEquals(CONTENT_TYPE, updatedBlob.contentType());
+    assertTrue(storage.delete(bucket, blobName));
+  }
+
+  @Test
+  public void testUpdateBlobFail() {
+    String blobName = "test-update-blob-fail";
+    BlobInfo blob = BlobInfo.of(bucket, blobName);
+    assertNotNull(storage.create(blob));
+    try {
+      storage.update(blob.toBuilder().contentType(CONTENT_TYPE).generation(42L).build(),
```

Comment:
As you said, it's not really possible that (meta)generation is 42: no tests share the same filenames and the bucket name is created using a random UUID. Anyway I agree, if -1 is an admissible value for the APIs that can never be the (meta)generation of an object then we should use it rather than 42.